src/smvlang/smv_range.h

```diff
@@ -58,32 +58,30 @@ class smv_ranget
     return from == to;
   }
 
-  smv_ranget &operator+(const smv_ranget &other)
+  smv_ranget operator+(const smv_ranget &other) const
   {
-    from += other.from;
-    to += other.to;
-    return *this;
+    return {from + other.from, to + other.to};
   }
 
-  smv_ranget &operator-(const smv_ranget &other)
+  smv_ranget operator-(const smv_ranget &other) const
   {
-    from -= other.from;
-    to -= other.to;
-    return *this;
+    return {from - other.from, to - other.to};
   }
 
-  smv_ranget &operator*(const smv_ranget &other)
+  smv_ranget operator*(const smv_ranget &other) const
   {
     mp_integer p1 = from * other.from;
     mp_integer p2 = from * other.to;
     mp_integer p3 = to * other.from;
     mp_integer p4 = to * other.to;
 
-    from = std::min(p1, std::min(p2, std::min(p3, p4)));
-    to = std::max(p1, std::max(p2, std::max(p3, p4)));
+    mp_integer from = std::min(p1, std::min(p2, std::min(p3, p4)));
+    mp_integer to = std::max(p1, std::max(p2, std::max(p3, p4)));
 
-    return *this;
+    return {std::move(from), std::move(to)};
   }
+
+  friend std::ostream &operator<<(std::ostream &, const smv_ranget &);
```

Comment:
What is the purpose of marking this a `friend` rather than putting the declaration directly in the global scope?

Reply:
Moved.